### PR TITLE
added new currency converter process according new design

### PR DESCRIPTION
**Pull request by dmitr2ish**

### CODE REVIEW

File: src/main/java/com/github/hey_world_team/currency_converter/model/chain/Order.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+@AllArgsConstructor
+@EqualsAndHashCode
+@ToString
+@Slf4j
+@Builder
+public class Order {
+    private long id;
+    private Trader trader;
+    private Trade trade;
+    private boolean isBuy;
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/model/chain/Order.java:16]: may be better to change toBuy and toSell

File: src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@ToString
+@Slf4j
+@Builder
+public class Trade {
+    private long id;
+    private String coinICC;
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java:18]: I dont understand what is coinICC;

File: src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    private CurrencyType currency;
+    private BigDecimal price;
+    private boolean isBuy;
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java:21]: We have same variables "isBuy" and  "isSell" here and in Order Class.
Let's delete one from them.

File: src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyService.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import com.github.hey_world_team.currency_converter.model.chain.Chain;
+import com.github.hey_world_team.currency_converter.service.processing.ConverterProcessor;
+
+import java.util.Currency;
+import java.util.List;
+
+public interface CurrencyService {
+
+    boolean currencyIsActual(Chain chain);
+    long saveNewPrices(BestChainRequestContext bestChainRequestContext);
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyService.java:13]: what is this method?

File: src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+@Service
+public class CurrencyServiceImpl implements CurrencyService{
+
+    @Override
+    public boolean currencyIsActual(Chain chain) {
+        return chain.isTcActual() && chain.isScActual();
+    }
+
+    @Override
+    public long saveNewPrices(BestChainRequestContext bestChainRequestContext) {
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyServiceImpl.java:16]: what is this method?

File: src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import com.github.hey_world_team.currency_converter.model.BestChainRequestContext;
+import com.github.hey_world_team.currency_converter.model.chain.Chain;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CurrencyServiceImpl implements CurrencyService{
+
+    @Override
+    public boolean currencyIsActual(Chain chain) {
+        return chain.isTcActual() && chain.isScActual();
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/currency/CurrencyServiceImpl.java:12]: I dont understand what is going on here? How does this checking work?

File: src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ChoosedExchange.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.github.hey_world_team.currency_converter.service.exchange;
+
+public enum ChoosedExchange {
+    ALL;
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ChoosedExchange.java:4]: How many stocks will we add here?

File: src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ExchangeServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+@Service
+public class ExchangeServiceImpl implements ExchangeService {
+
+    @Override
+    public boolean chainIsActual(Chain chain) {
+        return chain.isScToIccActual() && chain.isIccToTcActual();
+    }
+
+    @Override
+    public long saveNewOrderList(BestChainRequestContext bestChainRequestContext) {
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ExchangeServiceImpl.java:16]: For what is this method?

File: src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ExchangeServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.github.hey_world_team.currency_converter.model.BestChainRequestContext;
+import com.github.hey_world_team.currency_converter.model.chain.Chain;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ExchangeServiceImpl implements ExchangeService {
+
+    @Override
+    public boolean chainIsActual(Chain chain) {
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/exchange/ExchangeServiceImpl.java:11]: I dont understand this moment.

File: src/main/java/com/github/hey_world_team/currency_converter/service/processing/ConverterProcessor.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                if (!currencyService.currencyIsActual(chain)) {
+                    // TODO (rowCount = 0) retry
+                    long rowCount = currencyService.saveNewPrices(bestChainRequestContext);
+                }
+                if (!exchangeService.chainIsActual(chain)) {
+                    // TODO (rowCount = 0) retry
+                    long rowCount = exchangeService.saveNewOrderList(bestChainRequestContext);
+                }
+            }
+            return chainService.getActualChainByRequest(bestChainRequestContext);
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/processing/ConverterProcessor.java:50]: Can we return just "chain" ?

File: src/main/java/com/github/hey_world_team/currency_converter/service/validation/ValidationServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+public class ValidationServiceImpl implements ValidationService{
+
+    @Override
+    public boolean validateRequest(BestChainRequestContext requestContext) {
+        //TODO add more checks
+        return requestContext.getTc() != null && requestContext.getSc() != null;
+    }
+
+    @Override
+    public boolean chainIsActual(Chain chain) {
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/service/validation/ValidationServiceImpl.java:17]: I didnt get it

File: src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CurrencyConverterApi.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        @RequestParam(value = "sourceCurrency") String sc,
+        @RequestParam(value = "targetCurrency") String tc,
+        @RequestParam(value = "intermediateCryptoCoins", required = false) String icc) {
+        BestChainRequestContext requestContext
+            = BestChainRequestContext.builder()
+                                     .id(UUID.randomUUID())
+                                     .countOfChains(countOfChains)
+                                     .decency(decency)
+                                     .verified(verified)
+                                     .ce((ce != null && ChoosedExchange.contains(ce))
```
Comment (rassl95) [inline src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CurrencyConverterApi.java:62]: Let's change "ce" to "cex" like centralized exchange

**Review by rassl95 — CHANGES_REQUESTED** [review]
I need to have some call with you

File: src/main/java/com/github/hey_world_team/currency_converter/model/chain/Order.java (hunk 1 above)
Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/model/chain/Order.java:16]: yes it's a good idea, i will rename the field toBuy, and the method will be isToBuy

File: src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java (hunk 3 above)
Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/model/chain/Trade.java:21]: Yes, you're right

File: src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CurrencyConverterApi.java (hunk 12 above)
Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CurrencyConverterApi.java:62]: yeah, good point, I'm agree

**Review by rassl95 — APPROVED** [review]

**Pull request merged**